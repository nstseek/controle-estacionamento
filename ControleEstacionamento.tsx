import React from 'react';
import { View, TextInput, Text, Button, AsyncStorage, AsyncStorageStatic } from 'react-native';
import * as Permissions from 'expo-permissions';
import * as TaskManager from 'expo-task-manager';
import * as BackgroundFetch from 'expo-background-fetch';
import * as Expo from 'expo';
import { LocalNotification } from 'expo/build/Notifications/Notifications.types';
import { taskName } from './App';

interface ControleEstacionamentoState {
    saldo: number;
    saldoTemp: string;
    valor: number;
    valorTemp: string;
    permNotifications: boolean;
}

export enum PersistentStorage {
    saldo = 'saldo',
    valor = 'valor'
}

const setContainerStyle = {
    flex: 1,
    flexDirection: 'row',
    height: 50,
    width: 200
};

const inputStyle = {
    borderColor: '#555',
    borderWidth: 1,
    borderStyle: 'solid',
    borderRadius: 3
};

const saldoStyle = {
    fontSize: 70
};

const inputContainerStyle = {
    width: 200,
    height: 40,
    marginRight: 10
};

const buttonContainerStyle = {
    width: 80,
    height: 40
};

export default class ControleEstacionamento extends React.Component {
    state: ControleEstacionamentoState;
    asyncStorage: AsyncStorageStatic;
    expoPermissions: any;

    constructor(props) {
        super(props);
        this.asyncStorage = AsyncStorage;
        this.state = {
            saldo: 90,
            saldoTemp: `90`,
            valor: 4.5,
            valorTemp: `4.5`,
            permNotifications: false
        };
        this.asyncStorage.getItem(PersistentStorage.saldo, this.saveSaldo.bind(this));
        this.asyncStorage.getItem(PersistentStorage.valor, this.saveValor.bind(this));
        this.askNotifPermissions();
    }

    askNotifPermissions = async () => {
        this.expoPermissions = await Permissions.getAsync(Permissions.NOTIFICATIONS);
        if (this.expoPermissions.status !== 'granted') {
            Permissions.askAsync(Permissions.NOTIFICATIONS);
        }
    };

    sendNotification = async (title: string, body: string) => {
        const notification: LocalNotification = {
            title: title,
            body: body
        };
        Expo.Notifications.presentLocalNotificationAsync(notification);
    };

    render() {
        return (
            <View>
                <Text>Valor do estacionamento:</Text>
                <View style={setContainerStyle}>
                    <View style={inputContainerStyle}>
                        <TextInput
                            style={inputStyle}
                            value={this.state.valorTemp.toString()}
                            onChangeText={this.setValorTemp.bind(this)}
                        ></TextInput>
                    </View>
                    <View style={buttonContainerStyle}>
                        <Button title='Set' onPress={this.setValor.bind(this)}></Button>
                    </View>
                </View>
                <Text>Forçar saldo atual no cartão:</Text>
                <View style={setContainerStyle}>
                    <View style={inputContainerStyle}>
                        <TextInput
                            style={inputStyle}
                            value={this.state.saldoTemp.toString()}
                            onChangeText={this.setSaldoTemp.bind(this)}
                        ></TextInput>
                    </View>
                    <View style={buttonContainerStyle}>
                        <Button title='Set' onPress={this.setSaldo.bind(this)}></Button>
                    </View>
                </View>
                <Text>Valor da diaria:</Text>
                <Text>{this.state.valor}</Text>
                <Text>Saldo atual:</Text>
                <Text style={saldoStyle}>{this.state.saldo}</Text>
                <Button title='-' onPress={this.decreaseSaldo.bind(this)}></Button>
            </View>
        );
    }

    setSaldo() {
        this.asyncStorage.setItem(PersistentStorage.saldo, this.state.saldoTemp, this.errorFunc);
        this.sendNotification('Controle Estacionamento', 'Saldo salvo.');
        this.setState((previousState) => {
            return {
                ...previousState,
                saldo: Number(this.state.saldoTemp)
            };
        });
    }

    setValor() {
        this.asyncStorage.setItem(PersistentStorage.valor, this.state.valorTemp, this.errorFunc);
        this.sendNotification('Controle Estacionamento', 'Valor salvo.');
        this.setState((previousState) => {
            return {
                ...previousState,
                valor: Number(this.state.valorTemp)
            };
        });
    }

    setSaldoTemp(text: string) {
        this.setState((previousState) => {
            return {
                ...previousState,
                saldoTemp: text
            };
        });
    }

    setValorTemp(text: string) {
        this.setState((previousState) => {
            return {
                ...previousState,
                valorTemp: text
            };
        });
    }

    decreaseSaldo() {
        if (this.state.saldo >= this.state.valor) {
            this.setState((previousState) => {
                return {
                    ...previousState,
                    saldo: this.state.saldo - this.state.valor
                };
            });
        }
        this.asyncStorage.setItem(PersistentStorage.saldo, this.state.saldo.toString(), this.errorFunc);
    }

    errorFunc(error) {
        if (error) {
            console.log(error);
        }
    }

    saveValor(error, result) {
        if (!error && result) {
            this.setState((previousState) => {
                return {
                    ...previousState,
                    valor: Number(result)
                };
            });
        } else if (result === null) {
            this.asyncStorage.setItem(PersistentStorage.valor, this.state.valor.toString(), this.errorFunc);
        }
    }

    saveSaldo(error, result) {
        if (!error && result) {
            this.setState((previousState) => {
                return {
                    ...previousState,
                    saldo: Number(result)
                };
            });
        } else if (result === null) {
            this.asyncStorage.setItem(PersistentStorage.saldo, this.state.saldo.toString(), this.errorFunc);
        }
    }
}
