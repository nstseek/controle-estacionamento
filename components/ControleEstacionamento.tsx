import React from 'react';
import { View, TextInput, Text, Button, AsyncStorage, AsyncStorageStatic } from 'react-native';
import Styles from './ControleEstacionamento.styles';
import { Screens, PersistentStorage } from '../types';

interface ControleEstacionamentoState {
    saldo: number;
    saldoTemp: string;
    valor: number;
    valorTemp: string;
    ready: boolean;
}

interface ControleEstacionamentoProps {
    setTela(tela: Screens): any;
}

export default class ControleEstacionamento extends React.Component<
    ControleEstacionamentoProps,
    ControleEstacionamentoState
> {
    asyncStorage: AsyncStorageStatic;
    hisStackSize: number;

    constructor(props: ControleEstacionamentoProps) {
        super(props);
        this.asyncStorage = AsyncStorage;
        this.hisStackSize = -1;
        this.state = {
            saldo: 90,
            saldoTemp: `90`,
            valor: 4.5,
            valorTemp: `4.5`,
            ready: false
        };
        this.getDataFromStorage();
    }

    render() {
        return (
            <View>
                <Text>Valor do estacionamento:</Text>
                <View style={Styles.setContainerStyle}>
                    <View style={Styles.inputContainerStyle}>
                        <TextInput
                            style={Styles.inputStyle}
                            value={this.state.valorTemp.toString()}
                            onChangeText={this.setValorTemp.bind(this)}
                        ></TextInput>
                    </View>
                    <View style={Styles.buttonContainerStyle}>
                        <Button title='Set' onPress={this.setValor.bind(this)}></Button>
                    </View>
                </View>
                <Text>Forçar saldo atual no cartão:</Text>
                <View style={Styles.setContainerStyle}>
                    <View style={Styles.inputContainerStyle}>
                        <TextInput
                            style={Styles.inputStyle}
                            value={this.state.saldoTemp.toString()}
                            onChangeText={this.setSaldoTemp.bind(this)}
                        ></TextInput>
                    </View>
                    <View style={Styles.buttonContainerStyle}>
                        <Button title='Set' onPress={this.setSaldo.bind(this)}></Button>
                    </View>
                </View>
                <Text>Valor da diaria:</Text>
                <Text>{this.state.valor}</Text>
                <Text>Saldo atual:</Text>
                <Text style={Styles.saldoStyle}>{this.state.saldo}</Text>
                <View style={Styles.button}>
                    <Button title='Subtrair' disabled={!this.state.ready} onPress={this.decreaseSaldo.bind(this)} />
                </View>
                <View style={Styles.button}>
                    <Button
                        title='Ver histórico'
                        disabled={!this.state.ready}
                        onPress={() => this.props.setTela(Screens.HISTORICO_ESTACIONAMENTO)}
                    />
                </View>
            </View>
        );
    }

    getDataFromStorage = async () => {
        const saldo = await this.asyncStorage.getItem(PersistentStorage.saldo);
        const valor = await this.asyncStorage.getItem(PersistentStorage.valor);
        const hisStackSize = await this.asyncStorage.getItem(PersistentStorage.historicoStackSize);
        this.hisStackSize = hisStackSize ? Number(hisStackSize) : this.hisStackSize;
        this.setState((previousState) => ({
            ...previousState,
            valor: valor ? Number(valor) : previousState.valor,
            saldo: saldo ? Number(saldo) : previousState.saldo,
            ready: true
        }));
    };

    getHistoricoKey = (write: boolean) => PersistentStorage.historicoStack + (this.hisStackSize + (write ? 1 : 0));
    // se o stackSize for < 0, n pode ler pq n existe stack negativo ne
    // se for escrever, passa true como parametro e false caso va ler

    updateStackSize = (): Promise<null> =>
        new Promise(async (resolve) => {
            this.hisStackSize = this.hisStackSize + 1;
            await this.asyncStorage.setItem(
                PersistentStorage.historicoStackSize,
                this.hisStackSize.toString(),
                this.errorFunc
            );
            resolve(null);
        });

    writeToStack = (data: string): Promise<null> =>
        new Promise(async (resolve) => {
            await this.asyncStorage.setItem(this.getHistoricoKey(true), data, this.errorFunc);
            await this.updateStackSize();
            resolve(null);
        });

    formatDate = (date: Date) =>
        `${date.getDate() < 10 ? '0' + date.getDate() : date.getDate()}/${
            date.getMonth() + 1 < 10 ? '0' + (date.getMonth() + 1) : date.getMonth() + 1
        }/${date.getFullYear()}`;

    formatTime = (date: Date) =>
        `${date.getHours() < 10 ? '0' + date.getHours() : date.getHours()}:${
            date.getMinutes() < 10 ? '0' + date.getMinutes() : date.getMinutes()
        }:${date.getSeconds() < 10 ? '0' + date.getSeconds() : date.getSeconds()}`;

    setSaldo = async () => {
        await this.asyncStorage.setItem(PersistentStorage.saldo, this.state.saldoTemp, this.errorFunc);
        const date = new Date();
        await this.writeToStack(`${this.formatDate(date)}|Saldo - ${this.state.saldoTemp}|${this.formatTime(date)}`);
        this.setState((previousState) => {
            return {
                ...previousState,
                saldo: Number(this.state.saldoTemp)
            };
        });
    };

    setValor = async () => {
        this.asyncStorage.setItem(PersistentStorage.valor, this.state.valorTemp, this.errorFunc);
        const date = new Date();
        await this.writeToStack(`${this.formatDate(date)}|Valor - ${this.state.valorTemp}|${this.formatTime(date)}`);
        this.setState((previousState) => {
            return {
                ...previousState,
                valor: Number(this.state.valorTemp)
            };
        });
    };

    setSaldoTemp(text: string) {
        this.setState((previousState) => {
            return {
                ...previousState,
                saldoTemp: text
            };
        });
    }

    setValorTemp(text: string) {
        this.setState((previousState) => {
            return {
                ...previousState,
                valorTemp: text
            };
        });
    }

    decreaseSaldo = async () => {
        if (this.state.saldo >= this.state.valor) {
            const date = new Date();
            await this.asyncStorage.setItem(
                PersistentStorage.saldo,
                (this.state.saldo - this.state.valor).toString(),
                this.errorFunc
            );
            await this.writeToStack(`${this.formatDate(date)}|Decrease - ${this.state.valor}|${this.formatTime(date)}`);
            this.setState((previousState) => {
                return {
                    ...previousState,
                    saldo: this.state.saldo - this.state.valor
                };
            });
        }
    };

    errorFunc(error) {
        if (error) {
            throw new Error(error);
        }
    }
}
