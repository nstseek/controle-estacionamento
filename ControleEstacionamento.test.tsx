import React from 'react';
import ControleEstacionamento, { PersistentStorage } from './ControleEstacionamento';
import renderer from 'react-test-renderer';
import { AsyncStorage } from 'react-native';

const mockAsyncStorage = {
    ...AsyncStorage,
    getItem: jest.fn(),
    setItem: jest.fn()
}

it('renders correctly', () => {
    expect(renderer.create(<ControleEstacionamento />).toJSON()).toMatchSnapshot();
});
describe('testa funcoes', () => {
    it('deve chamar o asyncStorage pra salvar o saldo', () => {
        const component = new ControleEstacionamento(null);
        component.asyncStorage = mockAsyncStorage;
        component.decreaseSaldo();
        expect(mockAsyncStorage.setItem).toHaveBeenCalledWith(PersistentStorage.saldo, component.state.saldo.toString(), component.errorFunc);
    })
    it('deve chamar o asyncStorage pra salvar o saldo', () => {
        const component = new ControleEstacionamento(null);
        component.asyncStorage = mockAsyncStorage;
        component.setSaldo();
        expect(mockAsyncStorage.setItem).toHaveBeenLastCalledWith(PersistentStorage.saldo, component.state.saldoTemp, component.errorFunc);
    })
    it('deve chamar o asyncStorage pra salvar o valor', () => {
        const component = new ControleEstacionamento(null);
        component.asyncStorage = mockAsyncStorage;
        component.setValor();
        expect(mockAsyncStorage.setItem).toHaveBeenCalledWith(PersistentStorage.valor, component.state.valorTemp.toString(), component.errorFunc);
    })
});
