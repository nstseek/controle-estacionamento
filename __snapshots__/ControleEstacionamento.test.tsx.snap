// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <Text>
    Valor do estacionamento:
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "height": 50,
        "width": 200,
      }
    }
  >
    <View
      style={
        Object {
          "height": 40,
          "marginRight": 10,
          "width": 200,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "#555",
            "borderRadius": 3,
            "borderStyle": "solid",
            "borderWidth": 1,
          }
        }
        underlineColorAndroid="transparent"
        value="4.5"
      />
    </View>
    <View
      style={
        Object {
          "height": 40,
          "width": 80,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Set
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Text>
    Forçar saldo atual no cartão:
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "height": 50,
        "width": 200,
      }
    }
  >
    <View
      style={
        Object {
          "height": 40,
          "marginRight": 10,
          "width": 200,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "#555",
            "borderRadius": 3,
            "borderStyle": "solid",
            "borderWidth": 1,
          }
        }
        underlineColorAndroid="transparent"
        value="90"
      />
    </View>
    <View
      style={
        Object {
          "height": 40,
          "width": 80,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Set
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Text>
    Valor da diaria:
  </Text>
  <Text>
    4.5
  </Text>
  <Text>
    Saldo atual:
  </Text>
  <Text
    style={
      Object {
        "fontSize": 70,
      }
    }
  >
    90
  </Text>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        -
      </Text>
    </View>
  </View>
</View>
`;
